addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js

```diff
@@ -1283,11 +1283,301 @@ export class OdooEditor extends EventTarget {
+                // in this case the algorithm should prepend recovered nodes to this.editable
+                docStartRoot = this.editable;
+            }
+            if (docEnd === this.editable) {
+                // the selection removed the end fragment entirely from this.editable
+                // in this case the algorithm should append recovered nodes to this.editable
+                docEndRoot = this.editable;
+            }
+            if (docStart !== this.editable && docEnd !== this.editable) {
+                // recover the last child of this.editable present in this selection
+                // using the fact that the last element of contentsElements corresponds to docEnd or is one of its children
+                let docE = docEnd;
+                let fragE = contentsElements[contentsElements.length - 1];
+                while (fragE && fragE.oid && fragE.parentElement) {
+                    // fragE could have been fully selected (and thus is a child of docE)
+                    // fragE cannot have an oid if it is still present in the document.
+                    fragE = fragE.parentElement;
+                }
+                while (fragE && fragE.parentElement) {
+                    // docE and fragE are synchronized, reducing depth until fragE has no parent (its parent would then be this.editable)
+                    docE = docE.parentElement;
+                    fragE = fragE.parentElement;
+                }
+                docEndRoot = docE;
+                docStartRoot = docE.previousElementSibling || docEndRoot;
+            }
+
+            // extract children of root from an array of unremovables which is ordered and in reverse order.
+            // children of root are guaranteed to be at the end of unremovables if any.
+            // to maintain O(n) complexity, push and reverse are used instead of shift
+            const extractUnremovableChildren = function (root, unremovables) {
+                const unremovableChildren = [];
+                while (unremovables.length) {
+                    const unremovable = unremovables.pop();
+                    if (root.contains(unremovable)) {
+                        unremovableChildren.push(unremovable);
+                    } else {
+                        unremovables.push(unremovable);
+                        break;
+                    }
+                }
+                return unremovableChildren.reverse();
+            };
+
+            // used to fill nodes with an empty <p><br></p> (i.e. when emptying an contentEditable=True)
+            const ensureEmptyPElement = function (nodes) {
+                if (!nodes.length) {
+                    const pElement = this.document.createElement('P');
+                    const brElement = this.document.createElement('BR');
+                    pElement.append(brElement);
+                    nodes.push(pElement);
+                }
+            }.bind(this);
```

Comment:
I just realized there were a lot of those inline functions actually. If they do not use variables from outside their internal scope then I suggest moving them in utils. For this particular one which you used to `bind`, you could pass the correct `document` as argument for example.